lib/classes/ConfigSchemaHandler/normalizeAjvErrors.test.js

```diff
@@ -0,0 +1,462 @@
+'use strict';
+
+const expect = require('chai').expect;
+const { normalizeAjvErrors, getListOfErrorMessages } = require('./normalizeAjvErrors');
+
+describe('#normalizeAjvErrors', () => {
```

Comment:
Before we discuss implementation of `normalizeAjvErrors`, I think it'll be good to agree on shape of its tests.

As I already noted in other comments, I think it'll be nice if:
- It tests just exported `normalizeAjvErrors` function as black box (no `getListOfErrorMessages` direct testing)
- Generate AJV errors for injection directly via AJV, with prepared schema and service config (configured in scope of this test)

Things I think we should cover:
- Unsupported root level property (as e.g. `foo` directly on root)
- Unsupported second level property (as `package.incclude` below)
- Unsupported event (as `functions.someFunc.events[0].httpp` below)
- Unsupported event property, for event configured with simple object schema (as `functions.someFunc.events[1].http.invalidProp` below)
- Unsupported event property, for event configured with two possible value types:
  - Schema example: `{ name: "alb", schema: { anyOf: [{ type: "string" }, { type: "object", properties: { foo: { type: "string" } }, additionalProperties: false }] } }`,
  - Unsupported property example: `functions.someFunc.events[x].alb.invalidProp`

I think that might be enough for now. What do you think?

Reply:
So my understanding is that those tests would involve creating AJV instance to which schema and user config will be passed. This seems like a good idea.

I just want to make things clearer about the contents of `normalizeAjvErrors.js` file:
- `normalizeAjvErrors` is internal function that _takes_ an array of AJV error objects and _outputs_ another array of 'smart' error objects. The length of output array equals to number of errors that will be shown to a user.
- `getListOfErrorMessages` is used by `ConfigSchemaHandler` class. It also _takes_ a list of AJV errors and _outputs_ an array of strings which are error messages that will be shown to a user. Internally this function uses `normalizeAjvErrors` to filter out only relevant AJV errors. The job of this function is to build error _messages_ based on an array of normalized _objects_.

In light of that, I believe  `getListOfErrorMessages` needs to be exported. I may hide `normalizeAjvErrors` as internal but this would mean there would not be an example of normalized array of those 'smart' errors. However, I may add that example to a comment but may be quite long.

Reply:
I think ideally it should be as:

`normalizeAjvErrors` - takes errors as produced by AJV, and returns list of all errors that should be shown to users (that's it)

So all job happens in context of `normalizeAjvErrors` internals. They can be constructed with `getListOfErrorMessages` internally, but that should be treated as black box, and not exported outside.

I assume that you intended to export multiple blocks of `normalizeAjvErrors` logic to have better means of programmatical testing of errors (?) If it's the case, then I would probably construct returned errors so they are programmatically testable, e.g. final error (as returned by `normalizeAjvErrors` ) could have folllowing format:

```javascript
{
  message: "Unsupported function event 'alb' at functions.function.events[0]"
  dataPath: ".functions['function'].events[0]", // As on AJV error
  keyword: "should match some schema in anyOf" // As on AJV error
}
```

And in testing we may simply test just against `dataPath` and `keyword` (do not test messages). While CLI should just display all returned messages.

Reply:
>  I assume that you intended to export multiple blocks of normalizeAjvErrors logic to have better means of programmatical testing of errors (?) 

Yes, that was my intention. For the final error, I ended up with this structure:
```javascript
{
  message: "Unsupported function event 'alb' at functions.function.events[0]",
  error: { ... } // Error object
}
```